File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/HtmlUtils.java
package app.zxtune.fs;

import android.net.Uri;
import android.text.TextUtils;

import androidx.annotation.Nullable;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;

import app.zxtune.io.Io;

public class HtmlUtils {

  public static Document parseDoc(InputStream input) throws IOException {
    // Looks like jsoup works not well enough for streamed parsing when page size is more than
    // several kb
    return Jsoup.parse(readHtml(input), "");
  }

  private static String readHtml(InputStream input) throws IOException {
    try {
      final ByteBuffer buf = Io.readFrom(input);
      return new String(buf.array(), 0, buf.limit(), "UTF-8");
    } finally {
      input.close();
    }
  }

  @Nullable
  public static Integer getQueryInt(Element anchor, String param) {
    final String raw = Uri.parse(anchor.attr("href")).getQueryParameter(param);
    return tryGetInteger(raw);
  }

  @Nullable
  public static Integer tryGetInteger(@Nullable String raw) {
    if (TextUtils.isEmpty(raw)) {
      return null;
    }
    final String txt = raw.trim();
    return TextUtils.isEmpty(txt) || !TextUtils.isDigitsOnly(txt) ? null : Integer.valueOf(txt);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/HtmlUtils.kt
package app.zxtune.fs

import android.net.Uri
import app.zxtune.io.Io.readFrom
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.nodes.Node
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset

object HtmlUtils {
    @Throws(IOException::class)
    // Looks like jsoup works not well enough for streamed parsing when page size is more than
    // several kb
    @JvmStatic
    fun parseDoc(input: InputStream): Document = Jsoup.parse(readHtml(input), "")

    private fun readHtml(input: InputStream): String = input.use {
        with(readFrom(input)) {
            String(array(), 0, limit(), Charset.forName("UTF-8"))
        }
    }

    @JvmStatic
    fun getQueryInt(anchor: Element, param: String): Int? =
        Uri.parse(anchor.attr("href"))
            .getQueryParameter(param)?.let { tryGetInteger(it) }

    @JvmStatic
    fun tryGetInteger(raw: String?): Int? = raw?.trim()?.toIntOrNull()
}

object Jsoup {
    internal fun Element.findFirst(cssSelector: String): Element? = selectFirst(cssSelector)
    internal fun Element.findFirstText(cssSelector: String): String? =
        findFirst(cssSelector)?.text()

    internal val Element.parent: Element? get() = parent()
    internal val Element.nextSibling: Node? get() = nextSibling()
    internal val Element.nextElementSibling: Element? get() = nextElementSibling()
}
